### PR TITLE
Deduplicate "using config from" message

### LINKED ISSUE #10509: What should we report when there are duplicate imports?
What should we report when there are duplicate imports?

Is it acceptable that there are duplicates reported for configuration, such as the 5 times for `- yops/yops-9.config` with the project added in #10508?

```
$ cabal run cabal-install:exe:cabal -- build --dry-run \
  --project-file=cabal-testsuite/PackageTests/ConditionalAndImport/yops-0.project
Warning: this is a debug build of cabal-install with assertions enabled.
When using configuration from:
  - yops-0.project
  - yops-2.config
  - yops-4.config
  - yops-4.config
  - yops-6.config
  - yops-6.config
  - yops-6.config
  - yops-8.config
  - yops-8.config
  - yops-8.config
  - yops-8.config
  - yops/yops-1.config
  - yops/yops-3.config
  - yops/yops-3.config
  - yops/yops-5.config
  - yops/yops-5.config
  - yops/yops-5.config
  - yops/yops-7.config
  - yops/yops-7.config
  - yops/yops-7.config
  - yops/yops-7.config
  - yops/yops-9.config
  - yops/yops-9.config
  - yops/yops-9.config
  - yops/yops-9.config
  - yops/yops-9.config
The following errors occurred:
  - The package directory '.' does not contain any .cabal file.
```

#9933 would reject duplicate imports.

### PR DESCRIPTION
**Pull request by philderbeast**
Adds an `Ord` instance for `ProjectConfigPath` and fixes the duplication in the reported "Configuration is affected by the following files", fixes #10509 by using `docProjectConfigFiles` instead of `docProjectConfigPath`.

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [ ] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added.

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Review by ulysses4ever — APPROVED** [review]

**Comment by ulysses4ever** [comment]
> This fixes the duplication in the reported "Configuration is affected by the following files"

I think it fixes only part of the duplication problem in the snippet you reference. I.e. it dedups file names but it doesn't dedup the phrase `"Configuration is affected by the following files"`. The latter seem to be a bug in the way that snippet is written: there shouldn't be a `sequence_` at all. What it should look like is:

```haskell
    notice verbosity . render . vcat $
      text "Configuration is affected by the following files:"
        : [ text "-" <+> docProjectConfigPath path
          | Explicit path <- Set.toList $ projectConfigProvenance projectConfig
          ]

```

**Comment by ulysses4ever** [comment]
I'm already doing dedup of the intro message as a part of #10548 btw. It'd be best if this one only handled the Ord instance.

It's also a bit more complicated than I said above: with implicit config provenance, it'll print the intro message and then nothing else, which doesn't make sense. THis is not hard to fix of course.

**Comment by philderbeast** [comment]
@mpickering for these added tests do we already have a normalization of the project root directory for tests, something like `<PROJECT-ROOT>`, now that tests are run in temporary directories with #9717?

```diff
$ git diff 
...
--- a/cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs
+++ b/cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs
@@ -261,6 +261,49 @@ main = cabalTest . withRepo "repo" . recordMode RecordMarked $ do
   log "checking if we detect when the same config is imported via many different paths (we don't)"
   woopping <- cabal' "v2-build" [ "--project-file=woops-0.project" ]
 
+  log "checking \"using config from message\" without URI imports"
+  withDirectory "yops" $ do
+    yopping <- fails $ cabal' "v2-build" [ "--project-file=../yops-0.project" ]
+    assertOutputContains
+      (normalizeWindowsOutput "When using configuration from: \
+      \ - /tmp/cabal-testsuite-286573/yops-0.project \
+      \ - /tmp/cabal-testsuite-286573/yops-2.config \
+      \ - /tmp/cabal-testsuite-286573/yops-4.config \
+      \ - /tmp/cabal-testsuite-286573/yops-6.config \
+      \ - /tmp/cabal-testsuite-286573/yops-8.config \
+      \ - /tmp/cabal-testsuite-286573/yops/yops-1.config \
+      \ - /tmp/cabal-testsuite-286573/yops/yops-3.config \
+      \ - /tmp/cabal-testsuite-286573/yops/yops-5.config \
+      \ - /tmp/cabal-testsuite-286573/yops/yops-7.config \
+      \ - /tmp/cabal-testsuite-286573/yops/yops-9.config \
+      \ The following errors occurred: \
+      \  - The package directory '.' does not contain any .cabal file.")
+      yopping
+
+    return ()
+
+  log "checking \"using config from message\" with URI imports"
+  withDirectory "woops" $ do
+    woopping <- fails $ cabal' "v2-build" [ "--project-file=../woops-0.project" ]
+    assertOutputContains
+      (normalizeWindowsOutput "When using configuration from: \
+      \ - /tmp/cabal-testsuite-282695/woops-0.project \
+      \ - /tmp/cabal-testsuite-282695/woops-2.config \
+      \ - /tmp/cabal-testsuite-282695/woops-4.config \
+      \ - /tmp/cabal-testsuite-282695/woops-6.config \
+      \ - /tmp/cabal-testsuite-282695/woops-8.config \
+      \ - /tmp/cabal-testsuite-282695/woops/woops-1.config \
+      \ - /tmp/cabal-testsuite-282695/woops/woops-3.config \
+      \ - /tmp/cabal-testsuite-282695/woops/woops-5.config \
+      \ - /tmp/cabal-testsuite-282695/woops/woops-7.config \
+      \ - /tmp/cabal-testsuite-282695/woops/woops-9.config \
+      \ - https://www.stackage.org/lts-21.25/cabal.config \
+      \ The following errors occurred: \
+      \  - The package directory '.' does not contain any .cabal file.")
+      woopping
+
+    return ()
```

**Comment by ulysses4ever** [comment]
> something like `<PROJECT-ROOT>`

The previous person who bumped into this problem wasn't able to find an existing solution: https://github.com/haskell/cabal/pull/8617#issuecomment-1327994549

**Comment by philderbeast** [comment]
@ulysses4ever and @mpickering, to get around the `/tmp/cabal-testsuite-286573/yops-0.project` problem I used `assertRegex`.

I see that `cabal' "v2-build" [ "--project-file=yops-0.project" ]` does not set the test environment's project.

https://github.com/haskell/cabal/blob/7de199ad5eb91f8f32eee76ecd4577ac36c13225/cabal-testsuite/src/Test/Cabal/Monad.hs#L790-L791

**Comment by philderbeast** [comment]
This will need another review because previously I'd mucked up the `Ord` instance.

**Comment by philderbeast** [comment]
I didn't add a test of an http import importing another http config file.

**Comment by ulysses4ever** [comment]
I'm traveling this week, so will only be able to get to this next week at earliest. Sorry!

**Comment by philderbeast** [comment]
> I'm already doing dedup of the intro message

@ulysses4ever there are two similar messages. You've dealt with the `Configuration is affected by the following files` one in #10548 and I'm dealing with the other one here, with `When using configuration from`.

```
$ grep -R -E 'When using configuration from|Configuration is' ./**/*.hs --exclude='*.test.hs'
./cabal-install/src/Distribution/Client/ProjectConfig.hs:      "When using configuration from:\n"
./cabal-install/src/Distribution/Client/ProjectPlanning.hs:        text "Configuration is affected by the following files:" : configfiles
```

**Review by geekosaur — APPROVED** [review]

**Comment by geekosaur** [comment]
The CI rework (#10503) speeds CI up enormously, fwiw.

**Comment by philderbeast** [comment]
Thanks for the approvals @geekosaur and @ulysses4ever. I've got the CI failure for Windows sorted by putting one test into its own folder and avoiding the "Configuration is changed" failure from #10617. Other than squashing commits, I don't think I've made changes but it has been a month that this pull request has been sitting idle.

**Comment by ulysses4ever** [comment]
I didn't approve it because I didn't find time to properly review it, sorry. I'll try to but I'm sort of busy towards the end of the year, so maybe someone else will get to it before me...

**Comment by philderbeast** [comment]
That is odd, the changes approved panel shows that @ulysses4ever both approved and is a pending reviewer.

image

**Comment by geekosaur** [comment]
I think that's because he approved a very early (while it was still draft) version; check the timeline here. The most recent activity from him was a comment, though.

**Comment by philderbeast** [comment]
I guess we're in a kind of limbo, while @ulysses4ever has a pending re-review, I can neither request another review or invalidate his prior approval.
 
image

You're right @geekosaur, you're the only one to have reviewed the later changes since I fixed the `Ord` instance.

**Comment by geekosaur** [comment]
I can dismiss it (click the downarrow to the right of "2 approvals", then the 3-dots to the right of his approval), but I'm not sure if that's an admin action or something you can do. Do you want me to dismiss it?

**Comment by philderbeast** [comment]
Thanks @geekosaur, I can dismiss it.

**Review by Mikolaj — APPROVED** [review]
This is quite delicate and I guess the devil might be in the details, but the tests are solid and LGTM.

**Comment by philderbeast** [comment]
Label merge+no rebase is necessary when the pull request is from an organisation.

File: cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs
Comment (9999years) [inline cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs:261]: This test is already kind of huge and unreadable, so I'm not a fan of making it even bigger, but I recognize that you didn't write it. Let's mark this for cleanup in the future though.

**Review by 9999years — APPROVED** [review]
Looks reasonable. I think the release note could use a before/after example to show how this helps users. (I don't care about an `Ord` instance but I care a lot about error message quality!)

File: cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs
Comment (philderbeast) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs]: I've removed this test.

Comment (philderbeast) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs]: > Let's mark this for cleanup in the future though.

I can do this separately to this pull request.

Comment (philderbeast) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs]: > This test is already kind of huge and unreadable

I did write this test and wanted to exercise the import feature a little more. Maybe I went a bit overboard with it? The project import feature was poorly tested when I started working on it.

```
$ git checkout cabal-install-v3.10.1.0

$ grep -R -E 'import:' cabal-testsuite/**/*.project
cabal-testsuite/PackageTests/ConditionalAndImport/cabal-cyclical.project:import: cabal-cyclical.project
cabal-testsuite/PackageTests/ConditionalAndImport/cabal.project:import: extra.project
cabal-testsuite/PackageTests/RelativePathProjectImports/cabal.project:import: ./dep/cabal.project
```

**Comment by Kleidukos** [comment]
@philderbeast that label does not prevent merging, it's for after-merge :)

**Comment by 9999years** [comment]
(I had also approved it beforehand so as not to block the merge)

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.14

**Comment by ulysses4ever** [comment]
Hey @philderbeast ! In order to backport this to 3.14, we had to revert renaming `docProjectConfigPaths` → `docProjectConfigFiles` (only on the 3.14 branch). Is that okay with you?

**Comment by philderbeast** [comment]
@ulysses4ever I should be able to check on this tomorrow.

**Comment by philderbeast** [comment]
Yeah no problem there @ulysses4ever. This is to avoid an API change, isn't it, with https://github.com/haskell/cabal/commit/ce26c3c9ee76fce68e6180140e221eb5c99f2463?

**Comment by ulysses4ever** [comment]
@philderbeast correct. `master` stays with the names as in your original patch.

**Pull request merged**
